### PR TITLE
vm: fix unclaimedGas calculation

### LINKED ISSUE #3589: unclaimedGas execution result difference
Neo-go's execution results differ from C#

## Current Behavior
```json
{
    "id": 1,
    "jsonrpc": "2.0",
    "result": {
        "state": "FAULT",
        "gasconsumed": "231485",
        "script": "wh8MDGN1cnJlbnRJbmRleAwUvvIEMUA2KnfBUJnH5kwS9wC2ZdpBYn1bUgwU5THqcjqLpY5TkquLOLfxy/LIGN4SwB8MDHVuY2xhaW1lZEdhcwwU9WPqQLwoPU0OBcSOowWz8qBzQO9BYn1bUg==",
        "stack": [
            {
                "type": "Integer",
                "value": "4258708"
            },
            {
                "type": "ByteString",
                "value": "5THqcjqLpY5TkquLOLfxy/LIGN4="
            }
        ],
        "exception": "at instruction 120 (SYSCALL): can't calculate bonus of height unequal (BlockHeight + 1)",
        "notifications": []
    }
}
```

## Expected Behavior
from C#
```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
        "script": "wh8MDGN1cnJlbnRJbmRleAwUvvIEMUA2KnfBUJnH5kwS9wC2ZdpBYn1bUgwU5THqcjqLpY5TkquLOLfxy/LIGN4SwB8MDHVuY2xhaW1lZEdhcwwU9WPqQLwoPU0OBcSOowWz8qBzQO9BYn1bUg==",
        "state": "HALT",
        "gasconsumed": "694455",
        "exception": null,
        "notifications": [],
        "stack": [
            {
                "type": "Integer",
                "value": "0"
            }
        ]
    }
}
```

## Steps to Reproduce
```shell
curl --location 'https://rpc10.n3.nspcc.ru:10331' \
--header 'Content-Type: text/plain' \
--data '{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "invokescript",
  "params": ["wh8MDGN1cnJlbnRJbmRleAwUvvIEMUA2KnfBUJnH5kwS9wC2ZdpBYn1bUgwU5THqcjqLpY5TkquLOLfxy/LIGN4SwB8MDHVuY2xhaW1lZEdhcwwU9WPqQLwoPU0OBcSOowWz8qBzQO9BYn1bUg=="]
}'
```

### PR DESCRIPTION
**Pull request by AliceInHunterland**
Fix difference with C#:
```
(base) ekaterinapavlova@MacBook-Air-4 neo-go % curl -X POST
http://seed1t5.neo.org:20332 -H 'Content-Type: application/json' -d '{
  "jsonrpc": "2.0",
  "method": "getapplicationlog",
  "params":
  ["61681ce24dffea5481e9a50b10159b43b7ebfc21967b0b06fee7ff69c7123e3f"],
  "id": 1
}' | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time
  Current
                                 Dload  Upload   Total   Spent    Left
                                 Speed
100   429    0   269  100   160    582    346 --:--:-- --:--:-- --:--:--
   930
{
   "id" : 1,
   "jsonrpc" : "2.0",
   "result" : {
      "executions" : [
         {
            "exception" : null,
            "gasconsumed" : "198754",
            "notifications" : [],
            "stack" : [
               {
                  "type" : "Integer",
                  "value" : "0"
               }
            ],
            "trigger" : "Application",
            "vmstate" : "HALT"
         }
      ],
      "txid" :
      "0x61681ce24dffea5481e9a50b10159b43b7ebfc21967b0b06fee7ff69c7123e3
      f"
   }
}
```
```

(base) ekaterinapavlova@MacBook-Air-4 neo-go % curl -X POST  https://rpc .t5.n3.nspcc.ru:20331 -H 'Content-Type: application/json' -d '{
  "jsonrpc": "2.0",
  "method": "getapplicationlog",
  "params":
  ["61681ce24dffea5481e9a50b10159b43b7ebfc21967b0b06fee7ff69c7123e3f"],
  "id": 1
}' | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time
  Current
                                 Dload  Upload   Total   Spent    Left
                                 Speed
100   583  100   423  100   160   1424    538 --:--:-- --:--:-- --:--:--
  1969
{
   "id" : 1,
   "jsonrpc" : "2.0",
   "result" : {
      "executions" : [
         {
            "exception" : "at instruction 120 (SYSCALL): can't calculate
             bonus of height unequal (BlockHeight + 1)",
            "gasconsumed" : "198754",
            "notifications" : [],
            "stack" : [
               {
                  "type" : "Integer",
                  "value" : "4704605"
               },
               {
                  "type" : "ByteString",
                  "value" : "KfYYlDe/fxqqqm1yr7o5XLnQ7uk="
               }
            ],
            "trigger" : "Application",
            "vmstate" : "FAULT"
         }
      ],
      "txid" :
      "0x61681ce24dffea5481e9a50b10159b43b7ebfc21967b0b06fee7ff69c7123e3
      f"
   }
}

```
```
(base) ekaterinapavlova@MacBook-Air-4 neo-go % ./bin/neo-go contract invokefunction -r https://rpc.t5.n3.nspcc.ru:20331 -w ./testnet_wallet .json 1e6f88377a6c6bc4f683a5fc61eed5645ec5f123 unclaimedGas e9eed0b95c39baaf726daaaa1a7fbf379418f629 4704605
Enter account NWtk9HYWsf1njtSzA3XNgwZXRtriACcJ9G password > Warning: FAULT VM state returned from the RPC node: at instruction 120 (SYSCALL): can't calculate bonus of height unequal (BlockHeight + 1). Use --force flag to send the transaction anyway.
```

Close #3589

### CODE REVIEW

**Review by roman-khimov — COMMENTED** [review]
We need tests for it. IIUC the problem was just for zero (or nonexistent) balance? Regular balances behave differently (execution breaks), it's just a matter of check order?

File: pkg/core/native/native_neo.go [hunk 1]
```diff
@@ -750,18 +750,23 @@ func makeVoterKey(pub []byte, prealloc ...[]byte) []byte {
...
-		return nil, storage.ErrKeyNotFound
+		return big.NewInt(0), nil
 	}
 	st, err := state.NEOBalanceFromBytes(si)
 	if err != nil {
 		return nil, err
 	}
+	if st.Balance.Sign() == 0 {
+		return big.NewInt(0), nil
+	} else if st.Balance.Sign() < 0 {
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:764]: It's impossible, I'd remove it. This check is present in C# but in real life you can't have negative balance here.

**Review by AnnaShaleva — COMMENTED** [review]
The fix itself is legit, please add results for the T5 test with non-zero Neo balance and also please, add unit-tests to cover uncovered paths in `neo-go/pkg/core/native/native_test/neo_test.go`.

**Comment by AnnaShaleva** [comment]
>  it's just a matter of check order?

Yes, it is.

**Comment by roman-khimov** [comment]
I'd say that our old code was a bit better since it checked inputs first and only touched the DB afterwards. But it's not worth fixing on the C# side at the same time (introducing HF-dependent behaviour).

File: pkg/core/native/native_test/neo_test.go [hunk 2]
```diff
@@ -771,6 +771,36 @@ func TestNEO_CalculateBonus(t *testing.T) {
...
+	})
+
+	t.Run("unclaimed GAS", func(t *testing.T) {
+		amount := int64(100)
+		neoValidatorsInvoker.Invoke(t, true, "transfer", e.Validator.ScriptHash(), accH, amount, nil)
+
+		for range 10 {
+			e.AddNewBlock(t)
+		}
+		acc.Invoke(t, 550, "unclaimedGas", accH, e.Chain.BlockHeight()+1)
```
Comment (AliceInHunterland) [inline pkg/core/native/native_test/neo_test.go:799]: @AnnaShaleva, can the amount of unclaimed gas after 10 blocks be calculated (is there any formula)?

File: pkg/core/native/native_test/neo_test.go [hunk 3]
```diff
@@ -771,6 +771,36 @@ func TestNEO_CalculateBonus(t *testing.T) {
...
+	accH := acc.Signers[0].ScriptHash()
+
+	t.Run("non-existing account", func(t *testing.T) {
+		// non-existing account, should return zero unclaimed GAS.
+		acc.Invoke(t, 0, "unclaimedGas", util.Uint160{}, 1)
+	})
+
+	t.Run("zero balance", func(t *testing.T) {
+		acc.Invoke(t, 0, "unclaimedGas", accH, 1)
+	})
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:790]: These two cases are the same from the contract's PoW, because no account record exists for these accounts. Merge them into a single "non-existing account" case.

Comment (AnnaShaleva) [reply pkg/core/native/native_test/neo_test.go:790]: And "zero balance" case (L762, `if st.Balance.Sign() == 0 {`) is not tested at all, because to test it you need an existing account record with zero balance on it. Such record may be created by voting from account with zero NEO balance, vote for any registered candidate.

File: pkg/core/native/native_test/neo_test.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/core/native/native_test/neo_test.go:799]: Sure, it's exactly https://github.com/nspcc-dev/neo-go/blob/9a383608244a822772b9e6cea25846ddd2102b92/pkg/core/native/native_neo.go#L768

A part of reward is distributed for NEO holding and a part of reward is distributed for voting, but for your test-case it's only NEO holding. And see also the tail of `TestNEO_CalculateBonus`.

File: pkg/core/native/native_test/neo_test.go [hunk 4]
```diff
@@ -771,6 +771,36 @@ func TestNEO_CalculateBonus(t *testing.T) {
...
+
+	t.Run("unclaimed GAS", func(t *testing.T) {
+		amount := int64(100)
+		neoValidatorsInvoker.Invoke(t, true, "transfer", e.Validator.ScriptHash(), accH, amount, nil)
+
+		for range 10 {
+			e.AddNewBlock(t)
+		}
+		acc.Invoke(t, 550, "unclaimedGas", accH, e.Chain.BlockHeight()+1)
+		acc.InvokeFail(t, "can't calculate bonus of height unequal (BlockHeight + 1)", "unclaimedGas", accH, e.Chain.BlockHeight())
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:800]: We also need a successful case.

**Comment by AnnaShaleva** [comment]
> Regular balances behave differently (execution breaks)

> please add results for the T5 test with non-zero Neo balance

@AliceInHunterland, have you managed to invoke this method on T5 testnet for account with non-zero balance? Please, attach invocation results for both C# and Go nodes to the PR.

**Comment by AliceInHunterland** [comment]
For Go Testnet:
1. non-zero balance, valid height
```
(base) neo-go % ./bin/neo-go contract invokefunction -r https://rpc.t5.n3.nspcc.ru:20331 -w ./testnet_wallet.json a460e9333ad2d906645b7ced9bffd532c90ed54e unclaimedGas3 Nc6LJ79RodHzaz5BghHGChMZYRa9GqJvES
Enter account NWtk9HYWsf1njtSzA3XNgwZXRtriACcJ9G password > 
Network fee: 0.00037284
System fee: 0.00264301
Total fee: 0.00301585
Relay transaction (y|N)> y
ae8bc01212c9dcf01c73c1e8a3e60c334e37351358fd9b00061b989a866cc81e

(base) neo-go % curl -X POST https://rpc.t5.n3.nspcc.ru:20331 -H 'Content-Type: application/json' -d '{
  "jsonrpc": "2.0",
  "method": "getapplicationlog",
  "params": ["ae8bc01212c9dcf01c73c1e8a3e60c334e37351358fd9b00061b989a866cc81e"],
  "id": 1
}' | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   425  100   277  100   148    606    323 --:--:-- --:--:-- --:--:--   932
{
   "id" : 1,
   "jsonrpc" : "2.0",
   "result" : {
      "executions" : [
         {
            "exception" : null,
            "gasconsumed" : "264301",
            "notifications" : [],
            "stack" : [
               {
                  "type" : "Integer",
                  "value" : "14306775"
               }
            ],
            "trigger" : "Application",
            "vmstate" : "HALT"
         }
      ],
      "txid" : "0xae8bc01212c9dcf01c73c1e8a3e60c334e37351358fd9b00061b989a866cc81e"
   }
}
```
1. non-zero balance, invalid height
```
(base) neo-go % ./bin/neo-go contract invokefunction -r https://rpc.t5.n3.nspcc.ru:20331 -w ./testnet_wallet.json a460e9333ad2d906645b7ced9bffd532c90ed54e unclaimedGa
s2 Nc6LJ79RodHzaz5BghHGChMZYRa9GqJvES
Enter account NWtk9HYWsf1njtSzA3XNgwZXRtriACcJ9G password > 
Warning: FAULT VM state returned from the RPC node: at instruction 120 (SYSCALL): can't calculate bonus of height unequal (BlockHeight + 1).
Use --force flag to send the transaction anyway.

```

**Comment by AliceInHunterland** [comment]
For C# Testnet:

1. non-zero balance, valid height

```
neo-go % ./bin/neo-go contract invokefunction -r http://seed1t5.neo.org:20332 -w ./testnet_wallet.json a460e9333ad2d906645b7ced9bffd532c90ed54e unclaimedGas3 Nc6LJ79RodHzaz5BghHGChMZYRa9GqJvES

Enter account NWtk9HYWsf1njtSzA3XNgwZXRtriACcJ9G password > 
Network fee: 0.00037284
System fee: 0.00264301
Total fee: 0.00301585
Relay transaction (y|N)> y
a489ffdd2e3ae1f3c85e87106bb2571cec94058ba9d008565dd68a17aeebcc72

neo-go % curl -X POST http://seed1t5.neo.org:20332 -H 'Content-Type: application/json' -d '{
  "jsonrpc": "2.0",
  "method": "getapplicationlog",
  "params": ["a489ffdd2e3ae1f3c85e87106bb2571cec94058ba9d008565dd68a17aeebcc72"],
  "id": 1
}' | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   424    0   276  100   148    566    303 --:--:-- --:--:-- --:--:--   872
{
   "id" : 1,
   "jsonrpc" : "2.0",
   "result" : {
      "executions" : [
         {
            "exception" : null,
            "gasconsumed" : "264301",
            "notifications" : [],
            "stack" : [
               {
                  "type" : "Integer",
                  "value" : "14307350"
               }
            ],
            "trigger" : "Application",
            "vmstate" : "HALT"
         }
      ],
      "txid" : "0xa489ffdd2e3ae1f3c85e87106bb2571cec94058ba9d008565dd68a17aeebcc72"
   }
}
```
2. non-zero balance, invalid height
```
 neo-go % ./bin/neo-go contract invokefunction -r http://seed1t5.neo.org:20332 -w ./testnet_wallet.json a460e9333ad2d906645b7ced9bffd532c90ed54e unclaimedGas2 Nc6LJ79RodHzaz5BghHGChMZYRa9GqJvES
  
Enter account NWtk9HYWsf1njtSzA3XNgwZXRtriACcJ9G password > 
Warning: FAULT VM state returned from the RPC node: Specified argument was out of the range of valid values. (Parameter 'end').
Use --force flag to send the transaction anyway.

```

**Comment by AnnaShaleva** [comment]
@AliceInHunterland, we need results for both C# and Go nodes. For 1 application log from Go node is missing. For 2 transaction is not even sent, use `--force` flag to send it and attach C#/Go application logs.

Also, reviewcomments are not fixed.

File: pkg/core/native/native_test/neo_test.go (hunk 4 above)
Comment (AliceInHunterland) [reply pkg/core/native/native_test/neo_test.go:800]: `acc.Invoke(t, 550, "unclaimedGas", accH, e.Chain.BlockHeight()+1) ` - is a successful case, isn't it?

File: pkg/core/native/native_test/neo_test.go [hunk 5]
```diff
@@ -771,6 +771,43 @@ func TestNEO_CalculateBonus(t *testing.T) {
...
+	t.Run("unclaimed GAS", func(t *testing.T) {
+		amount := 100
+		defaultGASParBlock := 5
+		rewardDistance := 10
+		neoValidatorsInvoker.Invoke(t, true, "transfer", e.Validator.ScriptHash(), accH, amount, nil)
+
+		for range rewardDistance - 1 {
+			e.AddNewBlock(t)
+		}
+		expectedGas := int64(amount*rewardDistance*defaultGASParBlock) / int64(rewardDistance)
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:803]: It doesn't work like this. See https://github.com/nspcc-dev/neo-go/pull/3596 and https://github.com/nspcc-dev/neo-go/blob/9a383608244a822772b9e6cea25846ddd2102b92/pkg/core/native/native_neo.go#L768

**Review by AnnaShaleva — APPROVED** [review]
Good to go, we need 0.106.4...
